R/param-estimates.R

```diff
@@ -105,3 +108,71 @@ is_diag <- function(.name) {
 
   return(.ind[1] == .ind[2])
 }
+
+
+#' Add shrinkage details to param table
+#'
+#' Private helper to unpack shrinkage details and create a column on the input tibble
+#' with the shrinkage assigned to the relevant parameters.
+#' @param .param_df The parameter estimates table containing, at minimum `parameter_names` and `diag` columns.
+#' @param .summary A `bbi_nonmem_summary` object.
+#' @importFrom dplyr select filter left_join bind_rows
+#' @importFrom stringr str_detect
+#' @importFrom tibble add_column
+#' @keywords internal
+add_param_shrinkage <- function(.param_df, .summary) {
+
+  # extract shrinkage for final estimation method
+  shk <- .summary[[SUMMARY_SHRINKAGE]]
+  shk <- shk[[length(shk)]]
+
+  if (is.null(shk)) {
+    return(.param_df %>% add_column(!!SUMMARY_PARAM_SHRINKAGE := NA_real_))
+  }
+
+  # check for mixture model with multiple subpops
+  if (length(shk) != 1) {
+    warning(paste(
+      glue("When using `param_estimates()` with a mixture model (multiple subpops) the `{SUMMARY_PARAM_SHRINKAGE}` column will be all `NA`."),
+      glue("Users can manually extract shrinkage for each subpop from the `{SUMMARY_SHRINKAGE}` element of the `bbi_nonmem_summary` object."),
+      sep = "\n"
+      ))
+    return(.param_df %>% add_column(!!SUMMARY_PARAM_SHRINKAGE := NA_real_))
+  }
+
+  # select the first (and only) subpop
+  shk <- shk[[1]]
+
+  # filter to only omega and sigma diagonal elements
+  diag_df <- .param_df %>%
+                filter(.data[[SUMMARY_PARAM_DIAG]]) %>%
+                select({{ SUMMARY_PARAM_NAMES }}, {{ SUMMARY_PARAM_DIAG }})
+
+  # parse shrinkage for OMEGA diagonals
+  omega_df <- filter(diag_df, str_detect(.data[[SUMMARY_PARAM_NAMES]], "OMEGA"))
+  omega_shk <- shk[[SUMMARY_SHRINKAGE_OMEGA]]
+  if (nrow(omega_df) != length(omega_shk)) {
+    stop(paste(
+      glue("Found {nrow(omega_df)} OMEGA diagonals in parameter table and {length(omega_shk)} elements in `.summary[['{SUMMARY_SHRINKAGE}']][['{SUMMARY_SHRINKAGE_OMEGA}']]`."),
+      "Summary object may be malformed."
+      ))
+  }
+  omega_df[[SUMMARY_PARAM_SHRINKAGE]] <- omega_shk
+
+  # parse shrinkage for SIGMA diagonals
+  sigma_df <- filter(diag_df, str_detect(.data[[SUMMARY_PARAM_NAMES]], "SIGMA"))
```

Comment:
probably so, but this (like the proposed refactor) I'm going to skip in the interest of time. It may, in the end, make it more readable but I don't think it's worth the effort at this moment.